File: ios/GradeThread/Import/CSVParser.swift
import Foundation

/// US-667 — minimal CSV/TSV parser. A faithful Swift port of the web's
/// `src/lib/csv.ts` so a sheet that imports cleanly on the web imports the same
/// way on iOS: handles quoted fields, escaped quotes (`""`), and newlines
/// embedded inside quotes. Pure + deterministic → unit-tested directly.
enum CSVParser {

    enum Delimiter: String {
        case comma = ","
        case tab = "\t"
        case semicolon = ";"
        var character: Character { Character(rawValue) }
    }

    /// Strips a leading UTF-8 BOM (`﻿`) that Excel/Sheets "CSV UTF-8"
    /// exports prepend — `.whitespaces` trimming doesn't remove it, so without
    /// this the first header keeps the BOM and auto-mapping/delimiter detection
    /// see it glued to the first cell.
    static func stripBOM(_ s: String) -> String {
        s.hasPrefix("\u{FEFF}") ? String(s.dropFirst()) : s
    }

    /// Picks the delimiter from the first line by frequency. Considers tab,
    /// comma, and semicolon (European-locale exports use `;`); ties prefer comma,
    /// then tab, matching the prior comma-default behavior.
    static func detectDelimiter(_ sample: String) -> Delimiter {
        let firstLine = stripBOM(sample).split(separator: "\n", maxSplits: 1, omittingEmptySubsequences: false).first ?? ""
        let tabs = firstLine.filter { $0 == "\t" }.count
        let commas = firstLine.filter { $0 == "," }.count
        let semicolons = firstLine.filter { $0 == ";" }.count
        if commas >= tabs && commas >= semicolons { return .comma }
        if tabs >= semicolons { return .tab }
        return .semicolon
    }

    /// Parses delimited text into a grid of string cells.
    static func parse(_ input: String, delimiter: Delimiter) -> [[String]] {
        var rows: [[String]] = []
        var row: [String] = []
        var field = ""
        var inQuotes = false
        let delim = delimiter.character

        let chars = Array(input)
        var i = 0
        func pushField() { row.append(field); field = "" }
        func pushRow() { rows.append(row); row = [] }

        while i < chars.count {
            let ch = chars[i]
            if inQuotes {
                if ch == "\"" {
                    if i + 1 < chars.count, chars[i + 1] == "\"" {
                        field.append("\""); i += 1
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(ch)
                }
                i += 1
                continue
            }
            if ch == "\"" { inQuotes = true; i += 1; continue }
            if ch == delim { pushField(); i += 1; continue }
            if ch == "\r" { i += 1; continue }
            if ch == "\n" { pushField(); pushRow(); i += 1; continue }
            field.append(ch)
            i += 1
        }
        // Tail
        if !field.isEmpty || !row.isEmpty {
            pushField(); pushRow()
        }
        // Drop trailing fully-empty rows.
        while let last = rows.last, last.allSatisfy({ $0.trimmingCharacters(in: .whitespaces).isEmpty }) {
            rows.removeLast()
        }
        return rows
    }

    /// A parsed sheet block (input including the header row).
    struct Sheet: Equatable {
        let headers: [String]
        let rows: [[String]]
    }

    /// Parses input into `{ headers, rows }`, auto-detecting the delimiter and
    /// normalizing every row to the header width (padding short rows).
    static func parseSheet(_ input: String) -> Sheet {
        let clean = stripBOM(input)
        let delimiter = detectDelimiter(clean)
        var grid = parse(clean, delimiter: delimiter)
        guard !grid.isEmpty else { return Sheet(headers: [], rows: []) }
        let headers = grid.removeFirst().map { $0.trimmingCharacters(in: .whitespaces) }
        let width = headers.count
        let rows = grid.map { raw -> [String] in
            if raw.count == width { return raw }
            if raw.count > width { return Array(raw.prefix(width)) }
            return raw + Array(repeating: "", count: width - raw.count)
        }
        return Sheet(headers: headers, rows: rows)
    }
}

File: android/app/src/main/java/com/gradethread/app/importer/CsvParser.kt
package com.gradethread.app.importer

/**
 * US-1389 (iOS `CSVParser`, web `src/lib/csv.ts`): a minimal CSV/TSV parser.
 *
 * A faithful port rather than a library, so a sheet that imports cleanly on the
 * web imports identically here. Handles quoted fields, escaped quotes (`""`)
 * and newlines inside quotes — all three appear the moment anyone pastes a
 * description into a spreadsheet.
 */
object CsvParser {

    enum class Delimiter(val char: Char) { COMMA(','), TAB('\t'), SEMICOLON(';') }

    /**
     * Strips a leading UTF-8 BOM.
     *
     * Excel and Sheets prepend one to "CSV UTF-8" exports, and trimming
     * whitespace does not remove it — so without this the first header keeps the
     * BOM glued to it and auto-mapping silently fails on exactly the column that
     * matters most.
     *
     * US-2435: written as the ESCAPE, not the literal character. A literal BOM
     * here is invisible in every editor and diff, and Android lint's ByteOrderMark
     * check — which cannot tell deliberate data from an encoding accident —
     * failed the build on it. Same bytes matched, one of them readable.
     */
    fun stripBom(input: String): String = input.removePrefix("\uFEFF")

    /**
     * The delimiter, by frequency in the first line.
     *
     * Semicolon is included because European-locale exports use it. Ties prefer
     * comma, then tab.
     */
    fun detectDelimiter(sample: String): Delimiter {
        val firstLine = stripBom(sample).substringBefore('\n')
        val tabs = firstLine.count { it == '\t' }
        val commas = firstLine.count { it == ',' }
        val semis = firstLine.count { it == ';' }
        return when {
            commas >= tabs && commas >= semis -> Delimiter.COMMA
            tabs >= semis -> Delimiter.TAB
            else -> Delimiter.SEMICOLON
        }
    }

    fun parse(input: String, delimiter: Delimiter): List<List<String>> {
        val rows = mutableListOf<List<String>>()
        var row = mutableListOf<String>()
        val field = StringBuilder()
        var inQuotes = false
        var i = 0

        while (i < input.length) {
            val ch = input[i]
            if (inQuotes) {
                if (ch == '"') {
                    if (i + 1 < input.length && input[i + 1] == '"') {
                        field.append('"')
                        i++
                    } else {
                        inQuotes = false
                    }
                } else {
                    field.append(ch)
                }
                i++
                continue
            }
            when (ch) {
                '"' -> inQuotes = true
                delimiter.char -> {
                    row.add(field.toString()); field.clear()
                }
                // Dropped, not appended: a Windows export would otherwise leave
                // a carriage return on the end of every last column.
                '\r' -> Unit
                '\n' -> {
                    row.add(field.toString()); field.clear()
                    rows.add(row); row = mutableListOf()
                }
                else -> field.append(ch)
            }
            i++
        }
        if (field.isNotEmpty() || row.isNotEmpty()) {
            row.add(field.toString())
            rows.add(row)
        }
        // Trailing blank lines are near-universal in exports and would otherwise
        // each become an empty item.
        while (rows.isNotEmpty() && rows.last().all { it.isBlank() }) {
            rows.removeAt(rows.size - 1)
        }
        return rows
    }

    data class Sheet(val headers: List<String>, val rows: List<List<String>>)

    /**
     * Headers plus rows, every row normalized to the header width.
     *
     * Short rows are padded and long ones truncated, so a column index from the
     * mapping is always safe to read — a ragged export is the normal case, not
     * the exception.
     */
    fun parseSheet(input: String): Sheet {
        val clean = stripBom(input)
        val grid = parse(clean, detectDelimiter(clean))
        if (grid.isEmpty()) return Sheet(emptyList(), emptyList())

        val headers = grid.first().map { it.trim() }
        val width = headers.size
        val rows = grid.drop(1).map { raw ->
            when {
                raw.size == width -> raw
                raw.size > width -> raw.take(width)
                else -> raw + List(width - raw.size) { "" }
            }
        }
        return Sheet(headers, rows)
    }
}
